lgc/elfLinker/ElfLinker.cpp

```diff
@@ -0,0 +1,645 @@
+// Three ways this can exit:
+// 1. On success, returns true.
+// 2. Returns false on failure due to something in the shaders or pipeline state making separate
+//    compilation and linking impossible. The client typically then does a whole-pipeline
+//    compilation instead. The client can call Pipeline::getLastError() to get a textual representation of the
+//    error, for use in logging or in error reporting in a command-line utility.
+// 3. Other failures cause exit by report_fatal_error. The client can that catch by setting a diagnostic handler
+//    with LLVMContext::setDiagnosticHandler, although the usefulness of that is limited, as no attempt is
+//    made by LLVM to avoid memory leaks.
+//
+// @param [out] outStream : Stream to write linked ELF to
+// @return : True for success, false if something about the pipeline state stops linking
+bool ElfLinkerImpl::link(raw_pwrite_stream &outStream) {
+  // Initialize symbol table and string table
+  m_symbols.push_back({});
+  m_strings = std::string("", 1);
+  m_stringMap[""] = 0;
+  // Pre-create four fixed sections at the start:
+  // 0: unused (per ELF spec)
+  // 1: string table
+  // 2: symbol table
+  // 3: .text
+  // 4: .note
+  m_outputSections.push_back(OutputSection(this, "", ELF::SHT_NULL));
+  m_outputSections.push_back(OutputSection(this, ".strtab", ELF::SHT_STRTAB));
+  m_outputSections.push_back(OutputSection(this, ".symtab", ELF::SHT_SYMTAB));
+  unsigned textSectionIdx = m_outputSections.size();
+  m_outputSections.push_back(OutputSection(this, ".text"));
+  m_outputSections.push_back(OutputSection(this, ".note", ELF::SHT_NOTE));
+
+  // Allocate input sections to output sections, and gather and merge PAL metadata.
+  m_pipelineState->clearPalMetadata();
+  for (auto &elfInput : m_elfInputs) {
+    for (const object::SectionRef &section : elfInput.objectFile->sections()) {
+      object::ELFSectionRef elfSection(section);
+      if (elfSection.getType() == ELF::SHT_PROGBITS &&
+          (elfSection.getFlags() & (ELF::SHF_ALLOC | ELF::SHF_WRITE)) == ELF::SHF_ALLOC) {
+        // All text and rodata sections get lumped together, even if they have different names.
+        m_outputSections[textSectionIdx].addInputSection(elfInput, section);
+      } else if (elfSection.getType() != ELF::SHT_NOTE) {
+        // Put same-named sections together, and exclude .note.
+        StringRef name = cantFail(section.getName());
+        if (name.startswith(".AMDGPU.")) {
+          for (unsigned idx = 1;; ++idx) {
+            if (idx == m_outputSections.size()) {
+              m_outputSections.push_back(OutputSection(this));
+              m_outputSections[idx].addInputSection(elfInput, section);
+              break;
+            }
+            if (name == m_outputSections[idx].getName()) {
+              m_outputSections[idx].addInputSection(elfInput, section);
+              break;
+            }
+          }
+        }
```

Comment:
Why are we discarding sections whose names do not start with ".AMDGPU."?

Reply:
It was an attempt to exclude things like the symbol table and reloc sections. But I'll change it to do that in a better way...

Reply:
I thought I resolved this, but I haven't.

Reply:
Any resolution on this? The only thing "special" PROGBITS thing I see in a typical ELF is .AMDGPU.disasm. symtab and strtab aren't of type PROGBITS according to llvm-readelf.

Reply:
Yeah, sorry, I have a resolution for this locally (don't check for .AMDGPU at all, and just use PROGBITS), but I haven't got round to pushing it yet.

Reply:
Really done it this time.